core/src/light_protocol/query_handler.rs

```diff
@@ -0,0 +1,458 @@
+            next_request_id: RwLock::new(0),
+            peers: RwLock::new(HashSet::new()),
+            pending: RwLock::new(BTreeMap::new()),
+        }
+    }
+
+    // TODO(thegaram): share `send_message` and
+    // `send_message_with_throttling` with sync
+    fn send_message(
+        io: &NetworkContext, peer: PeerId, msg: &Message,
+        priority: SendQueuePriority,
+    ) -> Result<(), NetworkError>
+    {
+        QueryHandler::send_message_with_throttling(
+            io, peer, msg, priority, false,
+        )
+    }
+
+    fn send_message_with_throttling(
+        io: &NetworkContext, peer: PeerId, msg: &Message,
+        priority: SendQueuePriority, throttling_disabled: bool,
+    ) -> Result<(), NetworkError>
+    {
+        if !throttling_disabled && msg.is_size_sensitive() {
+            if let Err(e) = THROTTLING_SERVICE.read().check_throttling() {
+                debug!("Throttling failure: {:?}", e);
+                return Err(e);
+            }
+        }
+        let mut raw = Bytes::new();
+        raw.push(msg.msg_id().into());
+        raw.extend(msg.rlp_bytes().iter());
+        let _size = raw.len();
+
+        if let Err(e) = io.send(peer, raw, priority) {
+            debug!("Error sending message: {:?}", e);
+            return Err(e);
+        };
+
+        debug!(
+            "Send message({}) to {:?}",
+            msg.msg_id(),
+            io.get_peer_node_id(peer)
+        );
+        Ok(())
+    }
+
+    fn dispatch_message(
+        &self, io: &NetworkContext, peer: PeerId, msg_id: MsgId, rlp: Rlp,
+    ) -> Result<(), Error> {
+        trace!("Dispatching message: peer={:?}, msg_id={:?}", peer, msg_id);
+
+        match msg_id {
+            MsgId::GET_STATE_ROOT => self.on_get_state_root(io, peer, &rlp),
+            MsgId::STATE_ROOT => self.on_state_root(io, peer, &rlp),
+            MsgId::GET_STATE_ENTRY => self.on_get_state_entry(io, peer, &rlp),
+            MsgId::STATE_ENTRY => self.on_state_entry(io, peer, &rlp),
+            _ => {
+                warn!("Unknown message: peer={:?} msgid={:?}", peer, msg_id);
+                // io.disconnect_peer(peer, Some(UpdateNodeOperation::Remove));
+                Ok(())
+            }
+        }
+    }
+
+    fn handle_error(
+        &self, io: &NetworkContext, peer: PeerId, msg_id: MsgId, e: Error,
+    ) {
+        warn!(
+            "Error while handling message, peer={}, msg_id={:?}, error={:?}",
+            peer, msg_id, e
+        );
+
+        // TODO(thegaram): remove wildcard so that
+        // the compiler can help us cover all cases
+        let disconnect = match e.0 {
+            ErrorKind::InvalidStateRoot => true,
+            ErrorKind::InvalidProof => true,
+            ErrorKind::InvalidRequestId => true,
+            ErrorKind::PivotHashMismatch => false,
+            ErrorKind::InternalError => false,
+            _ => false,
+        };
+
+        if disconnect {
+            io.disconnect_peer(peer, None);
```

Comment:
For some kinds of error, need to update the underlying node table via 2nd parameter.